slapo/sharding/utils.py

```diff
@@ -49,26 +49,100 @@ def all_gather_along_dim(inp, dim, world_size, group):
     return ret
 
 
+def reduce_scatter_along_dim(inp, dim, world_size, group):
+    """reduce-scatter along the given dimension.
+
+    Paramters
+    ---------
+    inp: torch.Tensor
+        The input tensor to reduce-scatter.
+    dim: int
+        The dimension to all-gather along.
+    world_size: int
+        The number of processes in the group.
+    group: torch.distributed.ProcessGroup
+        The process group to all-gather.
+
+    Returns
+    -------
+    torch.Tensor
+        The reduce-scattered tensor.
+    """
+    # reduce_scatter always targets dim 0, so we transpose the target dim
+    # to dim 0, and transpose the result back.
+    assert inp.shape[dim] % world_size == 0, (
+        f"Reduce scatter dimension {dim} size {inp.shape} "
+        f"should be divisible by world size {world_size}"
+    )
+
+    temp = inp.transpose(0, dim) if dim != 0 else inp
+    temp = temp.contiguous()
+    scatter_shape = list(temp.shape)
+    scatter_shape[0] //= world_size
+    ret = torch.empty(scatter_shape, dtype=inp.dtype).to(inp.device)
+
+    dist.reduce_scatter_tensor(ret, temp, group=group)
+    if dim != 0:
+        ret = ret.transpose(0, dim).contiguous()
+    return ret
+
+
+def scatter_along_first_dim(inp, dim, world_size, group):
```

Comment:
It's not always first.
```suggestion
def scatter_along_dim(inp, dim, world_size, group):
```